R/env.R

```diff
@@ -0,0 +1,702 @@
+    base::assign(nms[[i]], dict[[i]], envir = env_)
+  }
+
+  env
+}
+#' @rdname env_assign
+#' @export
+env_define <- function(env = env_caller(), ...) {
+  env_bind(env, list(...))
+}
+
+#' Assign a promise to an environment.
+#'
+#' These functions let you create a promise in an environment. Such
+#' promises behave just like lazily evaluated arguments. They are
+#' evaluated whenever they are touched by code, but not when they are
+#' passed as arguments.
+#'
+#' @inheritParams env_assign
+#' @param expr An expression to capture for
+#'   \code{env_assign_lazily()}, or a captured expression (either
+#'   quoted or a formula) for the standard evaluation version
+#'   \code{env_assign_lazily_()}. This expression is used to create a
+#'   promise in \code{env}.
+#' @param eval_env The environment where the promise will be evaluated
+#'   when the promise gets forced. If \code{expr} is a formula, its
+#'   environment is used instead. If not a formula and \code{eval_env}
+#'   is not supplied, the promise is evaluated in the environment
+#'   where \code{env_assign_lazily()} (or the underscore version) was
+#'   called.
+#' @seealso \code{\link{env_assign}()}
+#' @export
+#' @examples
+#' env <- env_new()
+#' env_assign_lazily(env, "name", cat("forced!\n"))
+#' env$name
+#'
+#' # Use the standard evaluation version with quoted expressions:
+#' f <- ~message("forced!")
+#' env_assign_lazily_(env, "name2", f)
+#' env$name2
+env_assign_lazily <- function(env = env_caller(), nm, expr, eval_env = NULL) {
+  f <- as_quoted_f(substitute(expr), eval_env)
+  env_assign_lazily_(env, nm, f)
+}
+#' @rdname env_assign_lazily
+#' @export
+env_assign_lazily_ <- function(env = env_caller(), nm, expr, eval_env = NULL) {
+  f <- as_quoted_f(expr, eval_env)
+
+  args <- list(
+    x = nm,
+    value = f_rhs(f),
+    eval.env = f_env(f),
+    assign.env = rlang::env(env)
+  )
+  do.call("delayedAssign", args)
+}
+
+#' Bury bindings and define objects in new scope.
+#'
+#' \code{env_bury()} is like \code{env_bind()} but it creates the
+#' bindings in a new child environment. Note that this function does
+#' not work by side effect.
+#'
+#' @inheritParams env_bind
+#' @return An object associated with the new environment.
+#' @export
+#' @examples
+#' scope <- env_new(env_base(), list(a = 10))
+#' fn <- function() a
+#' env(fn) <- scope
+#'
+#' # fn() sees a = 10:
+#' fn()
+#'
+#' # env_bury() will bury the current scope of fn() behind a new
+#' # environment:
+#' fn <- env_bury(fn, list(a = 1000))
+#' fn()
+env_bury <- function(env = env_caller(), dict = list()) {
+  env_ <- rlang::env(env)
+  env_ <- new.env(parent = env_)
+
+  env_bind(env_, dict)
+  env_set(env, env_)
+}
+
+#' Remove bindings from an environment.
+#'
+#' \code{env_unbind()} is the complement of
+#' \code{\link{env_bind}()}. Like \code{env_has()}, it ignores the
+#' parent environments of \code{env}. On the other hand,
+#' \code{env_eliminate()} is akin to \code{env_sees()} and will track
```

Comment:
The pairs of names don't seem obviously symmetric to me - now I think an `inherits` argument would be better, but we should default it to `FALSE`.